cc-emr-notebook/aws_emr_notebook.ipynb

```diff
@@ -0,0 +1,610 @@
+    "        return (record.rec_type == 'metadata' and\n",
+    "                record.content_type == 'application/json')\n",
+    "\n",
+    "    @staticmethod\n",
+    "    def is_html(record):\n",
+    "        \"\"\"Return true if (detected) MIME type of a record is HTML\"\"\"\n",
+    "        html_types = ['text/html', 'application/xhtml+xml']\n",
+    "        if (('WARC-Identified-Payload-Type' in record.rec_headers) and\n",
+    "            (record.rec_headers['WARC-Identified-Payload-Type'] in\n",
+    "             html_types)):\n",
+    "            return True\n",
+    "        for html_type in html_types:\n",
+    "            if html_type in record.content_type:\n",
+    "                return True\n",
+    "        return False\n",
+    "\n",
+    "\n",
+    "class ServerCountJob(JupyterCCSparkJob):\n",
+    "    \"\"\" Count server names sent in HTTP response header\n",
+    "        (WARC and WAT is allowed as input)\"\"\"\n",
+    "\n",
+    "    name = \"CountServers\"\n",
+    "    fallback_server_name = '(no server in HTTP header)'\n",
+    "\n",
+    "    def process_record(self, record):\n",
+    "        server_name = None\n",
+    "\n",
+    "        if self.is_wat_json_record(record):\n",
+    "            # WAT (response) record\n",
+    "            record = json.loads(record.content_stream().read())\n",
+    "            try:\n",
+    "                payload = record['Envelope']['Payload-Metadata']\n",
+    "                if 'HTTP-Response-Metadata' in payload:\n",
+    "                    server_name = payload['HTTP-Response-Metadata'] \\\n",
+    "                                         ['Headers'] \\\n",
+    "                                         ['Server'] \\\n",
+    "                                         .strip()\n",
+    "                else:\n",
+    "                    # WAT request or metadata records\n",
+    "                    return\n",
+    "            except KeyError:\n",
+    "                pass\n",
+    "        elif record.rec_type == 'response':\n",
+    "            # WARC response record\n",
+    "            server_name = record.http_headers.get_header('server', None)\n",
+    "        else:\n",
+    "            # warcinfo, request, non-WAT metadata records\n",
+    "            return\n",
+    "\n",
+    "        if server_name and server_name != '':\n",
+    "            yield server_name, 1\n",
+    "        else:\n",
+    "            yield ServerCountJob.fallback_server_name, 1\n",
+    "\n",
+    "job = ServerCountJob()\n",
+    "job.run()"
+   ]
+  }
+ ],
+ "metadata": {
+  "kernelspec": {
+   "display_name": "PySpark",
+   "language": "",
+   "name": "pysparkkernel"
+  },
+  "language_info": {
+   "codemirror_mode": {
+    "name": "python",
+    "version": 2
+   },
+   "mimetype": "text/x-python",
+   "name": "pyspark",
+   "pygments_lexer": "python2"
+  }
+ },
+ "nbformat": 4,
+ "nbformat_minor": 4
+}
```

Comment:
Is this continued? At least, inspect the extracted web server names in the Parquet files. Of course, could also do some real data science and compare market share of web servers then and now (taking two WARC or WAT files).

Reply:
However, this would require some normalization, eg. strip versions.